File: Programmers/Algorithm_Kit/08_dfs_bfs/003.cpp
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

const int dy[] = { -1,0,1,0 };
const int dx[] = { 0,1,0,-1 };

int bfs(vector<vector<int>>& maps) {
    int n = maps.size();
    int m = maps[0].size();

    vector<vector<int>> visited(n, vector<int>(m, 0));

    visited[0][0] = 1;

    queue<pair<int, int>> q;
    q.push({ 0,0 });

    while (!q.empty()) {
        pair<int, int> cur = q.front(); q.pop();

        for (int i = 0; i < 4; i++) {
            int ny = cur.first + dy[i];
            int nx = cur.second + dx[i];

            if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue;

            if (maps[ny][nx] == 1 && !visited[ny][nx]) {
                visited[ny][nx] = visited[cur.first][cur.second] + 1;
                q.push({ ny,nx });
            }
        }
    }

    if (!visited[n - 1][m - 1]) {
        return -1;
    }
    else {
        return visited[n - 1][m - 1];
    }
}

int solution(vector<vector<int>> maps) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
    int answer = bfs(maps);

    return answer;
}

File: programmers/algorithmkit/08_dfs_bfs/003.java
import java.util.*;

class Solution {
    int[] dy = {-1, 0, 1, 0};
    int[] dx = {0, 1, 0, -1};

    public int solution(int[][] maps) {
        int answer = bfs(maps);

        return answer;
    }

    int bfs(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length; // bfs(n, m)으로 빼는 것이 정석이지만.. maps가 매개변수로 되어 있어서 그냥 가져왔다.

        int[][] visited = new int[n][m];

        visited[0][0] = 1;

        Queue<Pos> q = new ArrayDeque<>();
        q.offer(new Pos(0, 0));

        while (!q.isEmpty()) {
            Pos cur = q.poll();

            for (int i = 0; i < 4; i++) {
                int ny = cur.y + dy[i];
                int nx = cur.x + dx[i];

                if (ny < 0 || ny >= n || nx < 0 || nx >= m) {
                    continue;
                }

                // maps가 전역변수였다면 bfs(n, m) 원래 하던 방식으로 정의했을 것
                if (maps[ny][nx] == 1 && visited[ny][nx] == 0) {
                    visited[ny][nx] = visited[cur.y][cur.x] + 1;
                    q.offer(new Pos(ny, nx));
                }
            }
        }

        if (visited[n - 1][m - 1] == 0) {
            return -1;
        } else {
            return visited[n - 1][m - 1];
        }
    }

    // Pos, Pair, Info ...
    class Pos {
        int y;
        int x;

        public Pos(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
}
